### PR TITLE
Refactor and optimize Frame.where

### PR DESCRIPTION
**Pull request by vyasr**
This PR is a substantial refactoring of `Frame.where`. It removes many dead code paths, excises numerous unnecessary copies, and simplifies and consolidates various parts of the logic. It also splits up parts of the implementation into the specific Frame classes for which they are used. Prior to this PR, all the code was contained in a single function that essentially had completely independent code paths for DataFrame vs SingleColumnFrame. Splitting these into methods of the appropriate classes also makes mypy much happier.

The resulting code is significantly faster. I'll post more benchmarks soon, but we see improvements from 20% to up to 70%, even for reasonable data sizes (e.g. 1 million rows). You have to go past 10 million rows before the performance improvements are washed out by the sheer volume of computation time.

### CODE REVIEW

**Comment by vyasr** [comment]
<details><summary>Benchmarks</summary>
<p>


```
---------- benchmark '-alt0-series_dtype_int_rows_100-series_dtype_int_nulls_false_rows_100': 2 tests ----------
Name (time in us)                                                                                 Mean
----------------------------------------------------------------------------------------------------------------
bench_where[-alt0-series_dtype_int_rows_100-series_dtype_int_nulls_false_rows_100] (0001)     170.0649 (1.0)
bench_where[-alt0-series_dtype_int_rows_100-series_dtype_int_nulls_false_rows_100] (0002)     186.9814 (1.10)
----------------------------------------------------------------------------------------------------------------

---------- benchmark '-alt0-series_dtype_int_rows_100-series_dtype_int_nulls_true_rows_100': 2 tests ----------
Name (time in us)                                                                                Mean
---------------------------------------------------------------------------------------------------------------
bench_where[-alt0-series_dtype_int_rows_100-series_dtype_int_nulls_true_rows_100] (0001)     191.3864 (1.0)
bench_where[-alt0-series_dtype_int_rows_100-series_dtype_int_nulls_true_rows_100] (0002)     225.9981 (1.18)
---------------------------------------------------------------------------------------------------------------

---------- benchmark '-alt0-series_dtype_int_rows_10000-series_dtype_int_nulls_false_rows_10000': 2 tests ----------
Name (time in us)                                                                                     Mean
--------------------------------------------------------------------------------------------------------------------
bench_where[-alt0-series_dtype_int_rows_10000-series_dtype_int_nulls_false_rows_10000] (0001)     164.9918 (1.0)
bench_where[-alt0-series_dtype_int_rows_10000-series_dtype_int_nulls_false_rows_10000] (0002)     187.4164 (1.14)
--------------------------------------------------------------------------------------------------------------------

---------- benchmark '-alt0-series_dtype_int_rows_10000-series_dtype_int_nulls_true_rows_10000': 2 tests ----------
Name (time in us)                                                                                    Mean
-------------------------------------------------------------------------------------------------------------------
bench_where[-alt0-series_dtype_int_rows_10000-series_dtype_int_nulls_true_rows_10000] (0001)     208.1943 (1.0)
bench_where[-alt0-series_dtype_int_rows_10000-series_dtype_int_nulls_true_rows_10000] (0002)     228.9239 (1.10)
-------------------------------------------------------------------------------------------------------------------

---------- benchmark '-alt0-series_dtype_int_rows_1000000-series_dtype_int_nulls_false_rows_1000000': 2 tests ----------
Name (time in us)                                                                                         Mean
------------------------------------------------------------------------------------------------------------------------
bench_where[-alt0-series_dtype_int_rows_1000000-series_dtype_int_nulls_false_rows_1000000] (0001)     540.2130 (1.0)
bench_where[-alt0-series_dtype_int_rows_1000000-series_dtype_int_nulls_false_rows_1000000] (0002)     545.6433 (1.01)
------------------------------------------------------------------------------------------------------------------------

---------- benchmark '-alt0-series_dtype_int_rows_1000000-series_dtype_int_nulls_true_rows_1000000': 2 tests ----------
Name (time in us)                                                                                        Mean
-----------------------------------------------------------------------------------------------------------------------
bench_where[-alt0-series_dtype_int_rows_1000000-series_dtype_int_nulls_true_rows_1000000] (0001)     605.1137 (1.0)
bench_where[-alt0-series_dtype_int_rows_1000000-series_dtype_int_nulls_true_rows_1000000] (0002)     626.1591 (1.03)
-----------------------------------------------------------------------------------------------------------------------

---------- benchmark '-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_100-dataframe_dtype_int_nulls_false_cols_1_rows_100': 2 tests ----------
Name (time in us)                                                                                                                                Mean
---------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_100-dataframe_dtype_int_nulls_false_cols_1_rows_100] (0001)     195.9932 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_100-dataframe_dtype_int_nulls_false_cols_1_rows_100] (0002)     389.3515 (1.99)
---------------------------------------------------------------------------------------------------------------------------------------------------------------

---------- benchmark '-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_100-dataframe_dtype_int_nulls_true_cols_1_rows_100': 2 tests ----------
Name (time in us)                                                                                                                               Mean
--------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_100-dataframe_dtype_int_nulls_true_cols_1_rows_100] (0001)     226.8745 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_100-dataframe_dtype_int_nulls_true_cols_1_rows_100] (0002)     437.7565 (1.93)
--------------------------------------------------------------------------------------------------------------------------------------------------------------


---------- benchmark '-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_10000-dataframe_dtype_int_nulls_false_cols_1_rows_10000': 2 tests ----------
Name (time in us)                                                                                                                                    Mean
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_10000-dataframe_dtype_int_nulls_false_cols_1_rows_10000] (0001)     198.4275 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_10000-dataframe_dtype_int_nulls_false_cols_1_rows_10000] (0002)     393.6598 (1.98)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------

---------- benchmark '-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_10000-dataframe_dtype_int_nulls_true_cols_1_rows_10000': 2 tests ----------
Name (time in us)                                                                                                                                   Mean
------------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_10000-dataframe_dtype_int_nulls_true_cols_1_rows_10000] (0001)     241.4914 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_10000-dataframe_dtype_int_nulls_true_cols_1_rows_10000] (0002)     440.0462 (1.82)
------------------------------------------------------------------------------------------------------------------------------------------------------------------

---------- benchmark '-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_1000000-dataframe_dtype_int_nulls_false_cols_1_rows_1000000': 2 tests ----------
Name (time in us)                                                                                                                                        Mean
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_1000000-dataframe_dtype_int_nulls_false_cols_1_rows_1000000] (0001)     554.2396 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_1000000-dataframe_dtype_int_nulls_false_cols_1_rows_1000000] (0002)     748.8435 (1.35)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------

---------- benchmark '-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_1000000-dataframe_dtype_int_nulls_true_cols_1_rows_1000000': 2 tests ----------
Name (time in us)                                                                                                                                       Mean
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_1000000-dataframe_dtype_int_nulls_true_cols_1_rows_1000000] (0001)     642.6329 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_1-dataframe_dtype_int_cols_1_rows_1000000-dataframe_dtype_int_nulls_true_cols_1_rows_1000000] (0002)     847.9108 (1.32)
----------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------- benchmark '-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_100-dataframe_dtype_int_nulls_false_cols_6_rows_100': 2 tests -----------
Name (time in us)                                                                                                                                  Mean
-----------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_100-dataframe_dtype_int_nulls_false_cols_6_rows_100] (0001)       897.8136 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_100-dataframe_dtype_int_nulls_false_cols_6_rows_100] (0002)     1,751.7117 (1.95)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------

--------- benchmark '-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_100-dataframe_dtype_int_nulls_true_cols_6_rows_100': 2 tests ---------
Name (time in ms)                                                                                                                             Mean
------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_100-dataframe_dtype_int_nulls_true_cols_6_rows_100] (0001)     1.0551 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_100-dataframe_dtype_int_nulls_true_cols_6_rows_100] (0002)     2.0256 (1.92)
------------------------------------------------------------------------------------------------------------------------------------------------------------

----------- benchmark '-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_10000-dataframe_dtype_int_nulls_false_cols_6_rows_10000': 2 tests -----------
Name (time in us)                                                                                                                                      Mean
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_10000-dataframe_dtype_int_nulls_false_cols_6_rows_10000] (0001)       873.5480 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_10000-dataframe_dtype_int_nulls_false_cols_6_rows_10000] (0002)     1,744.0243 (2.00)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------

--------- benchmark '-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_10000-dataframe_dtype_int_nulls_true_cols_6_rows_10000': 2 tests ---------
Name (time in ms)                                                                                                                                 Mean
----------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_10000-dataframe_dtype_int_nulls_true_cols_6_rows_10000] (0001)     1.1181 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_10000-dataframe_dtype_int_nulls_true_cols_6_rows_10000] (0002)     2.0045 (1.79)
----------------------------------------------------------------------------------------------------------------------------------------------------------------
--------- benchmark '-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_1000000-dataframe_dtype_int_nulls_false_cols_6_rows_1000000': 2 tests ---------
Name (time in ms)                                                                                                                                      Mean
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_1000000-dataframe_dtype_int_nulls_false_cols_6_rows_1000000] (0001)     4.1683 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_1000000-dataframe_dtype_int_nulls_false_cols_6_rows_1000000] (0002)     5.0017 (1.20)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------

--------- benchmark '-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_1000000-dataframe_dtype_int_nulls_true_cols_6_rows_1000000': 2 tests ---------
Name (time in ms)                                                                                                                                     Mean
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_1000000-dataframe_dtype_int_nulls_true_cols_6_rows_1000000] (0001)     4.1174 (1.0)
bench_where[-alt1-dataframe_dtype_int_cols_6-dataframe_dtype_int_cols_6_rows_1000000-dataframe_dtype_int_nulls_true_cols_6_rows_1000000] (0002)     5.3112 (1.29)
--------------------------------------------------------------------------------------------------------------------------------------------------------------------

---------- benchmark '_case_1-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_100': 2 tests ----------
Name (time in us)                                                                                                          Mean
-----------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_100] (0001)     216.0383 (1.0)
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_100] (0002)     425.7766 (1.97)
-----------------------------------------------------------------------------------------------------------------------------------------

---------- benchmark '_case_1-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_10000': 2 tests ----------
Name (time in us)                                                                                                            Mean
-------------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_10000] (0001)     196.5695 (1.0)
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_10000] (0002)     400.5418 (2.04)
-------------------------------------------------------------------------------------------------------------------------------------------

---------- benchmark '_case_1-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_1000000': 2 tests ----------
Name (time in us)                                                                                                              Mean
---------------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_1000000] (0001)     558.3505 (1.0)
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_1000000] (0002)     743.6757 (1.33)
---------------------------------------------------------------------------------------------------------------------------------------------

----------- benchmark '_case_1-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_100': 2 tests -----------
Name (time in us)                                                                                                            Mean
-------------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_100] (0001)       905.7402 (1.0)
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_100] (0002)     1,716.4353 (1.90)
-------------------------------------------------------------------------------------------------------------------------------------------

----------- benchmark '_case_1-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_10000': 2 tests -----------
Name (time in us)                                                                                                              Mean
---------------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_10000] (0001)       895.3995 (1.0)
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_10000] (0002)     1,735.4196 (1.94)
---------------------------------------------------------------------------------------------------------------------------------------------

--------- benchmark '_case_1-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_1000000': 2 tests ---------
Name (time in ms)                                                                                                            Mean
-------------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_1000000] (0001)     3.7444 (1.0)
bench_where[_case_1-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_1000000] (0002)     4.9776 (1.33)
-------------------------------------------------------------------------------------------------------------------------------------------

---------- benchmark '_case_2-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_100': 2 tests ----------
Name (time in us)                                                                                                          Mean
-----------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_100] (0001)     217.2899 (1.0)
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_100] (0002)     478.5137 (2.20)
-----------------------------------------------------------------------------------------------------------------------------------------

---------- benchmark '_case_2-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_10000': 2 tests ----------
Name (time in us)                                                                                                            Mean
-------------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_10000] (0001)     211.9532 (1.0)
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_10000] (0002)     475.7557 (2.24)
-------------------------------------------------------------------------------------------------------------------------------------------

---------- benchmark '_case_2-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_1000000': 2 tests ----------
Name (time in us)                                                                                                              Mean
---------------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_1000000] (0001)     566.4992 (1.0)
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_1-dataframe_dtype_int_nulls_false_cols_1_rows_1000000] (0002)     834.2240 (1.47)
---------------------------------------------------------------------------------------------------------------------------------------------

----------- benchmark '_case_2-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_100': 2 tests -----------
Name (time in us)                                                                                                            Mean
-------------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_100] (0001)       930.0574 (1.0)
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_100] (0002)     2,095.2110 (2.25)
-------------------------------------------------------------------------------------------------------------------------------------------

----------- benchmark '_case_2-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_10000': 2 tests -----------
Name (time in us)                                                                                                              Mean
---------------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_10000] (0001)       947.9763 (1.0)
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_10000] (0002)     2,084.4131 (2.20)
---------------------------------------------------------------------------------------------------------------------------------------------

--------- benchmark '_case_2-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_1000000': 2 tests ---------
Name (time in ms)                                                                                                            Mean
-------------------------------------------------------------------------------------------------------------------------------------------
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_1000000] (0001)     4.1949 (1.0)
bench_where[_case_2-dataframe_dtype_int_nulls_false_cols_6-dataframe_dtype_int_nulls_false_cols_6_rows_1000000] (0002)     5.4645 (1.30)
-------------------------------------------------------------------------------------------------------------------------------------------

```
</p>
</details>

**Comment by vyasr** [comment]
@shwina @galipremsagar I could use a second look at a few of the TODO questions in the code. Ignore the TODOs that are clearly just action items ("need to do X"), there are a couple that are questions that I am not immediately sure how to answer.

**Comment by vyasr** [comment]
> is `0001` this PR and `0002` branch-22.08?

@galipremsagar Yes. This PR is the faster one :wink:

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -2606,6 +2606,90 @@ def set_index(
...
+            if cond_col := cond._data.get(name):
+                result = cudf._lib.copying.copy_if_else(
+                    col, other_col, cond_col
+                )
+
+                out[name] = _make_categorical_like(result, self._data[name])
+            else:
+                # TODO: Can we reuse masks here? All columns will be of the
+                # same length, but unsure if there are any cases where the
+                # reference semantics will become problematic later on.
```
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py:2682]: I would have thought so, but I'm not sure what sequence of operations to perform to trigger that from a high-level API. Maybe it has to be done at the lowest level of the Cython level? For instance, this example seems safe:
```
>>> import cudf
>>> s = cudf.Series([1, None, 3])
>>> s2 = cudf.Series([1, 2, None])
>>> s2._column
s2._column        s2._column_names  s2._columns       
>>> s3 = s2._column.set_mask(s._column.mask)
>>> s[1] = 10
>>> s3  # Looks like s3 still has the original mask from s, not the updated mask
<cudf.core.column.numerical.NumericalColumn object at 0x7f20ba260f40>
[
  1,
  null,
  0
]
dtype: int64
```

File: python/cudf/cudf/core/dataframe.py [hunk 2]
```diff
@@ -2606,6 +2606,90 @@ def set_index(
...
+                )
+        elif (
+            hasattr(cond, "__array_interface__")
+            and cond.__array_interface__["shape"] != self.shape
+        ):
+            raise ValueError("conditional must be same shape as self")
+        elif not isinstance(cond, DataFrame):
+            # TODO: Can we just use cudf.DataFrame(...) directly? I don't think
+            # we want to support any objects that _have_ to go through the
+            # pd.DataFrame constructor.
```
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py:2635]: OK cool will update then.

**Comment by vyasr** [comment]
rerun tests

**Review by galipremsagar — APPROVED** [review]
Nice Perf improvements Vyas! 🔥

File: python/cudf/cudf/core/dataframe.py [hunk 3]
```diff
@@ -2606,6 +2606,85 @@ def set_index(
...
+            if cond_col := cond._data.get(name):
+                result = cudf._lib.copying.copy_if_else(
+                    col, other_col, cond_col
+                )
+
+                out[name] = _make_categorical_like(result, self._data[name])
+            else:
+                # TODO: Can we reuse masks here? All columns will be of the
+                # same length, but unsure if there are any cases where the
+                # reference semantics will become problematic later on.
```
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:2677]:
```suggestion
```

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**